dds/DCPS/RTPS/Spdp.cpp

```diff
@@ -915,7 +915,23 @@ Spdp::handle_handshake_message(const DDS::Security::ParticipantStatelessMessage&
   DCPS::RepoId reader = src_participant;
   reader.entityId = ENTITYID_P2P_BUILTIN_PARTICIPANT_STATELESS_READER;
 
-  if (iter->second.auth_state_ == DCPS::AS_HANDSHAKE_REPLY && msg.related_message_identity.source_guid == GUID_UNKNOWN) {
+  if ((iter->second.auth_state_ == DCPS::AS_HANDSHAKE_REPLY || iter->second.auth_state_ == DCPS::AS_HANDSHAKE_REPLY_SENT) && msg.related_message_identity.source_guid == GUID_UNKNOWN) {
+    // The initiator will retry using the same request.  If the
+    // initiator retries very quickly and we change the response,
+    // i.e., different diffie, then we will never authenticate.  Thus,
+    // if the initiator sends the same request, then we should send
+    // the same response.
+    if (iter->second.has_last_stateless_msg_ &&
+        iter->second.last_handshake_request_data_ == msg.message_data[0]) {
```

Comment:
Make sure that `message_data` is nonempty before using it.

Reply:
Line 891 has a check that msg.message_data.length() != 0.